### ISSUE TITLE
Chapter 5 possible errata and doubt

### ISSUE THREAD

**Issue by LorenaGdL** [OpenCVBlueprints/OpenCVBlueprints#8]
Hi guys! As the title says, this is about Chapter 5 of the book, so this is mainly for you @StevenPuttemans 

**1. Reporting a possible trivial errata**. The cascade classification process in detail > Step2  > First paragraph: [...] *especially when knowing that a model of 24x24 pixels can yield more than 16,000 features* -> Shouldn't it be 160,000 features? Not the most important errata at all, but just letting you know.

**2. A question about the explanation of the maxFalseAlarmRate param**. I'm not sure if this is an errata, a confusing explanation, or just that I'm plain wrong. I'll start by copying the book explanation:

*- maxFalseAlarmRate: This is the threshold that defines how much of your negative samples need to be classified as negatives before the boosting process should stop adding weak classifiers to the current stage. The default value is 0.5 and ensures that a stage of weak classifier will only do slightly better than random guessing on the negative samples. Increasing this value too much could lead to a single stage that already filters out most of your given windows, resulting in a very slow model at detection time due to the vast amount of features that need to be validated for each window. This will simply remove the large advantage of the concept of early window rejection.*

So, what's confusing me is the second half the of paragraph. You say increasing the param value results in more filtering and more number of features to validate per stage. As I understand it, increasing the maxFalseAlarmRate implies letting the stage to misclassify more negative samples as positive ones. Therefore, less negative samples are rejected in that stage, and a higher number of accepted windows passes to the next stage. Also, I think (though I'm not too sure here) that the number of weak classifiers/features per stage is lower as the maxFalseAlarmRate is higher. In fact, as you say later on, *This process continues until the false acceptance ratio on the negative samples is lower than the maximum false alarm rate set.* Therefore, with a higher maxFalseAlarmRate, the chances to get a lower acceptance ratio increases, leading to stages with less features.

To sum up, did you really want to say *increasing this value* instead of *decreasing this value*, or did I simply get it wrong?

**3. An additional doubt (I can post it in the OpenCV forum if you prefer)**. Based on your experience, is there any con (at detection time) of using more number of stages, but with less restrictive conditions? e.g. using 20 stages with minHit 0.990 instead of using 10 stages with minHit 0.995 (supposing you have enough data to tackle training in both cases, and that the final minHit is the same -haven't done math-)

**4. Good job on the chapter ;)**

**Comment by StevenPuttemans** [comment]
@LorenaGdL thank you for the detailed report. I will go through it step by step and get back to you if it needs some further digging.

1. Should indeed be `160.000` thanks for reporting. I will add to the errata page and return it back to Packt Publishing also so that in future versions of the book, this can get fixed.

2. I am not sure ... let me put this into example values for the `maxFalseAlarmRate` parameter as how I understanded it.
  2.1 A value of 0.5 means 50% chance of a negative sample being classified as a negative and thus rejected by the current stage. If the value gets higher than this, more than 50% of the samples get classified as negAND thus you get for example 52% of the samples which is better than random guessing.
  2.2  A value of 0.7 means 70% chance of a negative sample being classified as a negative and thus rejected by the current stage. This means that more windows get classified correctly and thus a stage is harsher on the samples. Vice versa for 0.3 value.
  2.3 However, this all starts from the point that the parameter describes _This is the threshold that defines how much of your negative samples need to be classified as negatives before the boosting process should stop adding weak classifiers to the current stage._, which I am now considering possible wrong and should be _This is the threshold that defines how much of your negative samples **are allowed to be classified as positives** before the boosting process should stop adding weak classifiers to the current stage._ This would indeed make sense looking at the parameters name.
  2.4 Only way to be 100% sure is to run some tests on monday as I will do!
3. Using more stages will simply lead to a longer processing time per window and thus depending on your resolution this could explode exponentially. However the number of features are INDEED in relation to this. I am perfectly sure that you could make a 5 stage classifier with 50 features each, which works as fasts as a 50 stage classifier with 5 features each (carefully and manually designed of course). This is the difference between the Viola and Jones approach and the later and newer ICF/ACF approach. There they increase it with the stage, but not linearly but exponentially... so not really sure if there is a correct answer here. Only tests specific to the case would result in a conclusive answer I guess.

Thank you for the input! Do get back to me when needed!

**Comment by LorenaGdL** [comment]
You're always welcome ;) And, thank you also for your insights about point 3.

Coming back to 2, I'm glad to are considering that subtle change in the sentence, because that's exactly what I originally understood (though of course I may be wrong anyway). With such change it's the way-around, a value of 0.7 would mean sort of 70% chances of classify a negative sample as positive, so a much permissive stage.

Just to show an example that I think supports this theory, let's take a look at this training output:
N --- HR --- FA
1 ---- 1 ----- 1
2 ---- 1 ----- 0.9
3 ---- 1 ----- 0.7
4 ---- 1 ----- 0.5

If `maxFalseAlarmRate` is 0.5, the stage would have ended with 4 weak classifiers. If it were 0.7, it would have ended with just 3 weak classifiers, which would be a less harsh stage, right?

I'll wait for your tests, and I'll also try to take a look at the code to see if I can discover anything

Thank you very very much for everything! I'm here for whatever you need :)

**Comment by StevenPuttemans** [comment]
@LorenaGdL your explanation is making 150% sense and it is indeed all about subtleties here. I will add the erratum also to the page and report it back to Packt to avoid future misunderstanding of this parameter. 

Besides that I have taken the time to surf the net a bit and it seems to me that I just wrote the explanation in a wrong way, though your meaning was exactly what I was aiming for!

**Comment by StevenPuttemans** [comment]
As you can see, adaptations are made and this can be closed down. Thanks again for the report!

**Comment by LorenaGdL** [comment]
:+1: :smile:
